### PR TITLE
Refactor: Significant speed-up for ManaOptions

### LINKED ISSUE #7710: Too much mana calculations on multiple Cascading Cataracts (game freeze)
1x [[Cascading Cataracts]] card can slow down the game, 3x cards can freeze it.

Search problem code by `TODO: research too many manas and freeze`  (fbb08ab1a82bac3311cc47b3f74606ad25f17fbb).

Just put the card to battlefield to reproduce game freeze:
```
battlefield:Human:Cascading Cataracts:1
```

Related #7649

### PR DESCRIPTION
**Pull request by Alex-Vasile**
# Benchmark Results: 1,000x faster and 175x less RAM allocation
PRE: https://github.com/magefree/mage/commit/f99da274cdac3c566a2e5d6b7fcb986ddf2dfca1
POST: latest commit to this branch
Execution time of `testCascadingCataracts`: `510s -> 0.50s`
Memory usage of `addManaWithCosts()`: `20,220 MB -> 115 MB`

To get times, run *all* of `ManaOptionsTest` so that the startup time is not being counted as part of `testCascadingCataracts`.

Results are from a 16GB M1 MacBook Pro. Difference will likely become even more pronounced on computers with slower RAM and lower RAM bandwith.

# Changes
- Change `ManaOption` from an `ArrayList` to a `LinkedHastList` in order to speed up computations by not adding duplicates in the first place.
- Refactored `subtractCostAddMana` to not use be significantly more efficient, biggest change is that it no longer converts mana to string for comparison.
- Changed `getPossiblePayCombinations` to use a `LinkedHashSet` rather than a `List` in order to speed up computations by not adding duplicates in the first place.
- Removed `assertDuplicatedManaOptions` and `removeDuplicated` since they are not needed anymore.

# TODO
- [x] Answer and address all TODO comments
- [x] Figure out why .contains() isn't working as expected.
- [x] Add `hashCode()` and `equals()` to both ConditionalMana
- [x] Re-implement `getPossiblePayCombinations` to no longer use String to decrease memory requirement (It accounts for use 97% of `subtractCostAddMana` memory usage.
https://github.com/magefree/mage/blob/e2c7db0a47ea62b854c4333737a45a0b61e05897/Mage/src/main/java/mage/abilities/mana/ManaOptions.java#L494-L522
- [x] Figure out why isExistingManaCombination is using `Mana.getMoreValuableMana`. (Mana.getMoreValuableMana accounts for 99% of `subtractCostAddMana` CPU usage.
https://github.com/magefree/mage/blob/e2c7db0a47ea62b854c4333737a45a0b61e05897/Mage/src/main/java/mage/abilities/mana/ManaOptions.java#L536-L544
- [x] Figure out failing tests
- [x] Figure out how to write a reproducible benchmark test to calculate before and after. (currently connecting to server and adding problem cards (Cascading Cataracts, even post #9068 it will slow down the game if you add 5+ as well as 5 of each land under the same player)

Potential improvements on this
- Memoizing getPossiblePayCombinations since it gets called multiple times a round and accounts for the vast majority of the computation time. Much of that can be eliminated at the cost of slightly higher memory usage on the server.
- Look see if `removeFullyIncludedVariations` can be removed as it still does not scale well if players have large boards. The scaling issue caused by it should not show up the the course of most games.

Closes #7710.

### CODE REVIEW

File: Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTestUtils.java
Comment (Alex-Vasile) [inline Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTestUtils.java:19]: TODO

File: Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTest.java
Comment (Alex-Vasile) [inline Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTest.java:381]: Why did this pass below. Shouldn't the available options be {B}{U} (tapping basic lands) and {B} (paying {B}{U} to tap Cabal Coffers and get a single {B} back)?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:606]: This portion of the code did more than remove duplicates. It also removes the variations that are already included. E.g. if {B}{U} and {B} are in `this`, then {B} will be removed.

Is this a performance optimization or a requirement for how other parts of the game work?

File: Mage/src/main/java/mage/Mana.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/Mana.java:1145]: This seems like a mistake. Why is this adding generic and colourless to return the amount of colourless available.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:401]: ?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:464]: ?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:479]: ?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:530]: I still don't understand the need for this function

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:554]: Still need to fix

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:553]: If it's a performance optimization, it may not be worth the computational cost of doing this operation on a Set

**Comment by Alex-Vasile** [comment]
I added a test, but I think the performance difference is so great that the test is irrelevant.

With the current state of the PR, the following is only slightly laggy:
Many

Note the x25 Cascading Cataracts at the bottom. The current implementation of ManaOptions still freezes for several minutes with only 5 of each.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [reply Mage/src/main/java/mage/abilities/mana/ManaOptions.java]: This scales part of the algorithm scales *ingredibly* poorly, `O(n^2)`.

Does anyone know if this is an actual requirement for other parts of the code? As in, is the code assuming that if ManaOptions has {B}{U} then it will *not* have {B} or {U}. Or the reverse, if it has {B} then {B}{U} is not available?

If it's an optimization, I am going to disable it since I'm not convinced it's worth the computation cost.

The tests (other than those which compare the size of manaOptions to a hardcoded value) all seem to pass. I will try playing some games against the AI with it disabled and see if I find anything broken.

**Comment by Alex-Vasile** [comment]
Below is the scaling for a variation of `testCascadingCataracts`. Comment out the `Assert.asserEquals` line since that has a hardcoded number. Then replace the card numbers with `n` so that the same number of each card is added.

n | manaOptions.size()
-|---------
4 |	 3,698
5 |	11,146
6 | 	28,289
7 |	63,343

At these larger sizes `removeFullyIncludedVariations` starts becoming the bottleneck in terms of CPU again. The RAM usage is dominated by Mana.copy() and adding to the HashSet, neither of which are really avoidable at this point since I avoided copying or creating a new Mana object whenever I could.

There is not much to be done about the memory allocation since all of those Mana objects are needed. But disabling `removeFullyIncludedVariations` *significantly* speeds up the run time at these larger sizes.

**Comment by Alex-Vasile** [comment]
I think to get better performance/scaling (certainly for the memory aspect) we'd have to switch our mana calcs to the way that humans do it: starting with the abilities we want to activate first.

Pick an ability we want to activate, search for sources that can satisfy the mana requirements and stop looking as soon as we find the first one that does.

Now, I have a feeling this will scale better than trying to generate every possible combination of mana, but I don't know what kind of problem we would get instead, and if we can avoid infinite loops. Perhaps something to look into at some point in the future if the current approach becomes a bottleneck.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [reply Mage/src/main/java/mage/abilities/mana/ManaOptions.java]: Looking through code, it looks like an optimization. I am removing it since it scales very poorly, and the performance penalty of not doing it appears to be less than the performance penalty of keeping it in.

File: Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java
Comment (DeepCrimson) [inline Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java:87]: What's the reason for commenting out this assertion and other similar "mana variations don't fit" assertions below?

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/ConditionalMana.java:238]: Second sentence is very confusing imo due to referencing a variable named `that`.

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/ConditionalMana.java:242]: The variable names `o` and `that` don't do much to clarify what these variables are used for imo.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:3]: Could you find a more widely available exception type to use?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:18]: Is there a more pertinent place to put this comment, or perhaps more context you can provide? It's unclear how this StackOverflow question about HashSet.remove is relevant.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:596]: Overridden

**Comment by DeepCrimson** [comment]
Great work finding and implementing this huge performance win! 🎉

File: Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java
Comment (Alex-Vasile) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java:87]: This number in the assertion is based on applying `removeFullyIncludedVariations`. The function was an optimization to remove variations that were already accounted for (E.g. if both {R} and {R}{W} were present then {R} is removed since {R}{W} will already allow the activation of all abilities that {R} does).

Ultimately it was a performance optimization, but in benchmarking it, the function accounts for the large majority of the computational cost of generating the mana combinations. Far more than the amount I found that it saves by removing the combinations.

So without the optimization, the number will necessarily be different. I've left them there rather than disable them incase anyone goes back at some point to implement a more efficient algorithm.

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (Alex-Vasile) [reply Mage/src/main/java/mage/ConditionalMana.java:242]: Changed `o` to `obj`.
`that` is used as a natural sounding comparison between two values `this` vs `that`.

**Comment by Alex-Vasile** [comment]
@DeepCrimson. Did you take a look at the rest of the changes here (particularly in ManaOptions) and see anything strange with the actual algorithm changes?

File: Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java
Comment (DeepCrimson) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java:87]: Where is `removeFullyIncludedVariations` defined? I'm not able to find it in the codebase.

File: Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTest.java
Comment (DeepCrimson) [inline Mage.Tests/src/test/java/org/mage/test/utils/ManaOptionsTest.java:605]: It'd be helpful to add a comment block here explaining the calculation that brings the number of potential mana combinations to 6,902.

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (DeepCrimson) [reply Mage/src/main/java/mage/ConditionalMana.java:242]: Personally I'd name them something like `obj` -> `objectToCheck` and `that` -> `conditionalManaToCheck` but that's just my opinion. Curious what others think here.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:393]: *True

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:423]: You can replace this assignment with `return true;`. This will skip the remaining iterations of the loop without changing this function's return value. This will also let you get rid of the variable `oldManaWasReplaced` altogether.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:446]: *If there are no generic costs

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [inline Mage/src/main/java/mage/abilities/mana/ManaOptions.java:576]: Maybe name method something more unique like `getManaAtIndex` rather than adding the `Deprecated` tag to a method that isn't deprecated?

**Comment by DeepCrimson** [comment]
> @DeepCrimson. Did you take a look at the rest of the changes here (particularly in ManaOptions) and see anything strange with the actual algorithm changes?

Yes I took a look, left a few comments but changes look good to me! I'd approve this if I had merge access.

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (JayDi85) [inline Mage/src/main/java/mage/ConditionalMana.java:237]: Must override `hashCode` and `equals` for inherited `ConditionalMana` classes with extra fields. See example:
shot_220712_103027

`ConditionalMana` must have docs about override requirements (see above).
shot_220712_111210

File: Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java
Comment (JayDi85) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java:87]: @DeepCrimson I think it's about `removeDuplicated` and `getMoreValuableMana` usage -- it must replace any generic option with colored option. So alex removed that optimization due performance issue with Cascading Cataracts (no code usage - no freezes). It must be research -- is it compatible with AI and auto-payment (is there any code that depends on the colors inside the ManaOptions).

shot_220712_112029

Comment (JayDi85) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/DoublingCubeTest.java:87]: @Alex-Vasile must specify the exceptional case in the assert comment (why it was disabled).

**Review by JayDi85 — CHANGES_REQUESTED** [review]
1. Equal/hash fixes in inherited classes;
2. Research color depended code in ManaOptions usage (is it exists);
3. Additional test for ConditionalMana;

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (JayDi85) [reply Mage/src/main/java/mage/ConditionalMana.java:237]: Well, I checked all inherited `ConditionalMana`, no inner fields. So it's fine. But must contains docs about it anyway.

File: Mage/src/main/java/mage/Mana.java
Comment (JayDi85) [reply Mage/src/main/java/mage/Mana.java:1129]: Yes, it can be the error (added with improved colorless mana support, see 782190bac37fa104505e73e9dfe99a6d69b03c26). As I can see the pattern (search code by `switch (manaType) {`):
- mana class uses both generic and colorless types;
- mana pool class uses only colorless type.

Current code in the mana so it must use both generic and colorless types as diff types, see other methods like `set` and `increase`.

File: Mage/src/main/java/mage/ConditionalMana.java
Comment (Alex-Vasile) [reply Mage/src/main/java/mage/ConditionalMana.java:237]: I will go through and check them. Also, I feel like ConditionalManaBuilder needs a better constructor. There are dozens of implementations of it which implement the exact same thing (a build with a single instance of a ConditionalMana object and a rule string)

Comment (Alex-Vasile) [reply Mage/src/main/java/mage/ConditionalMana.java:237]: The equals and hashCode I will handle most of the rest inside #9225.

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (Alex-Vasile) [reply Mage/src/main/java/mage/abilities/mana/ManaOptions.java:423]: I will have to think about that one and see. My gut feeling is that I can't just return true since the code must generate all combinations until the criteria for the while loop is met. The change of state is equally important to the return value (possibly more so).

**Comment by Alex-Vasile** [comment]
I still need to look into this one, but this for loop also looks like an "optimization" and is *incredibly* expensive as the size of ManaOptions increases.
https://github.com/magefree/mage/blob/5e26066b24227ba4ca8159e9c6f4290656bdbc87/Mage/src/main/java/mage/abilities/mana/ManaOptions.java#L67-L77

**Comment by JayDi85** [comment]
Possible research/idea: use standard lib to work with a combinations data (see google's guava lib).

**Comment by JayDi85** [comment]
> Use Gauava for which part?

Possible mana's combinations/collections. It's can generate combinations, store it, sorting, deduplicating, etc. And it's already optimized for performance and memory usage.

File: Mage.Tests/src/test/java/org/mage/test/cards/mana/conditional/ConditionalManaTest.java
Comment (Alex-Vasile) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/conditional/ConditionalManaTest.java:335]: Is the test I added what you had in mind?

Comment (JayDi85) [reply Mage.Tests/src/test/java/org/mage/test/cards/mana/conditional/ConditionalManaTest.java:335]: Yes

**Comment by Alex-Vasile** [comment]
> 1. ~~Equal/hash fixes in inherited classes~~, docs about it.
> 
> 2. Research color depended code in ManaOptions usage (is it exists);
> 
> 3. Additional test for ConditionalMana;

I think 3 is satisfied. I don't know what you mean by 2.

**Comment by Alex-Vasile** [comment]
> Possible research/idea: use standard lib to work with a combinations data (see google's guava lib).

We can open a separate issue into this (seeing if there is a library that can make this faster). 

But I think this PR will solve the issue well enough for now.

**Comment by Alex-Vasile** [comment]
> * ~Equal/hash fixes in inherited classes~, docs about it.
> * Research color depended code in ManaOptions usage (is it exists);
> * Additional test for ConditionalMana;

I plan to merge this soon and move on.
What do you mean by "Research color depended code in ManaOptions usage (is it exists);"?

File: Mage/src/main/java/mage/abilities/mana/ManaOptions.java
Comment (DeepCrimson) [reply Mage/src/main/java/mage/abilities/mana/ManaOptions.java:423]: `oldManaWasReplaced` is initialized with a value of `false` on line 393. In a nested `for` loop, it's conditionally set to `true` on line 423. Its value is ultimately returned on line 433 with no potential to be flipped back to `false` if it's been set to `true`.

Does this method have effects on the state of the game, or is it just used to do a `true`/`false` check? If the method's purpose is just for checking if a value's `true`/`false`, line 423 can be replaced with `return true;` and line 433 can be replaced with `return false;`.

**Comment by JayDi85** [comment]
@Alex-Vasile do not merge active PRs until active discussion/research finish (it's a bad thing and goes to unstable/unfinished code and technical debt, see your use case with auto-choices in #9206). Also do not merge any UI/UX related PRs until review.

> What do you mean by "Research color depended code in ManaOptions usage (is it exists);"?

You must take ManaOptions objects and research the usage. Are there any code that:
- uses a color searching in ManaOptions (you added "any" instead multicolors -- is it broke something or not -- good unit tests aren't a proof due low coverage percentage, must be code review);
- uses color priority for AI calculation and mana payments. As example: mana payment uses color priority (pays colored mana first).

File: Mage/src/main/java/mage/Mana.java
Comment (Alex-Vasile) [inline Mage/src/main/java/mage/Mana.java:1395]: Hashing for Mana is going to overflow for large color values. Each mana color is allowed to go as high as `Integer.MAX_VALUE`.

Perhaps something from: https://stackoverflow.com/questions/4045063/how-should-i-map-long-to-int-in-hashcode

**Comment by Alex-Vasile** [comment]
Okay. I think I have addressed all of the concerns from before.

Instead of removing `getPossiblePayCombinations` I tried implementing a more efficient version of it but ultimately couldn't get it working properly. I scrapped it and just went back to the O(n^2) approach that we had before. Even with it's poor scaling, the board will have to get quiet large before it it becomes noticeable.

The benchmark listed at the top, 115MB of ram used (not retained all at once) and 0.5s, is for the following board:
- 5x Plain
- 5x Island
- 5x Swamp
- 5x Mountain
- 5x Forest
- 5x Desert
- 3x Cascading Cataracts

That's 33 lands total for 1 player. I don't think that many lands or mana sources will ever show up in most games. And if they do, I think that other portions of the engine (specifically the copying of state over and over) will be the bottleneck.

**Review by theelk801 — APPROVED** [review]

**Comment by Alex-Vasile** [comment]
I plan to merge this at the end of the week unless any issues can be found with it

**Review by JayDi85 — APPROVED** [review]

**Pull request merged**
